### ISSUE TITLE
Idiom for extensible string printing for TensorImpl subclasses

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#73506]
### 🐛 Describe the bug

https://github.com/pytorch/pytorch/pull/72881/files#r813335202

@cpuhrsch tells us that for out of tree TensorImpl extensions, there is no good way to override string printing behavior. Ordinarily, if you were to subclass a Tensor, you could simply override `__repr__` and `__str__` to provide your desired behavior. However, for a TensorImpl subclass, by default these present in Python as just plain Tensor, so you don't have the subclass interposition point, and the printing function `_tensor_str` doesn't have any provision for extensibility in this way.

It's probably worth explicitly specifying that if you're cool enough to have an allocated DispatchKey in core, you're also cool enough to have a hook in `_tensor_str` for dynamic dispatching your printing algorithm somewhere else. Another thought is that you can use the XLA tensor style hook to override which subclass your tensor presents as, see https://github.com/pytorch/pytorch/blob/b213041df304c05980b3209dddf1b595699c8b74/torch/csrc/autograd/python_variable.cpp#L164 

### Versions

master

**Comment by bdhirsh** [comment]
Another reason that this might be useful: functorch [handles it by monkey-patching `torch._tensor_str._str`](https://github.com/pytorch/functorch/blob/1f363e545b5be651755c89ea73a88074ddbf5b9d/functorch/__init__.py#L128) when you `import functorch`, which means that you might run into trouble if you want to import any other libraries that monkey-patch tensor printing.

There was a PR a few weeks ago to do the same thing in XLA (monkey-patch `_tensor_str._str` for xla tensors to do fewer unnecessary compilations), but I pointed out that that wouldn't play well with functorch, so they ended up moving it in core ([discussion](https://github.com/pytorch/xla/pull/3253#discussion_r778466121)).

**Comment by zou3519** [comment]
Yes please! functorch (and other external libraries) should not be in the business of monkey-patching `torch._tensor_str.str`

**Comment by cpuhrsch** [comment]
This would be tremendous!

**Comment by cpuhrsch** [comment]
Also note that this would be very useful for COO and CSR Tensor as well. There is ugly [one-off logic](https://github.com/pytorch/pytorch/blob/master/torch/_tensor_str.py#L330-L368) right now. This would be much nicer to have in a separate file etc.
